ui/desktop/app/routes/scopes/scope/projects.js

```diff
@@ -76,41 +70,45 @@ export default class ScopesScopeProjectsRoute extends Route {
   }
 
   willDestroy() {
-    this.job?.stop();
-    super.willDestroy();
+    super.willDestroy(...arguments);
+    this.poller.cancelAll();
   }
 
   /**
    * Poll for new sessions with credentials. Sends a notification for each new session that has a credential.
    */
-  async poller() {
+  poller = restartableTask(async () => {
     let sessions;
 
-    try {
-      sessions = await this.clientAgentSessions.getNewSessionsWithCredentials();
-    } catch (e) {
-      // If we're unauthenticated, try and re-authenticate
-      if (e.statusCode === 401 || e.statusCode === 403) {
-        const sessionData = this.session.data?.authenticated;
-        const auth_token_id = sessionData?.id;
-        const token = sessionData?.token;
-
-        try {
-          await this.ipc.invoke('addTokenToDaemons', {
-            tokenId: auth_token_id,
-            token,
-          });
-          this.job.start();
-          __electronLog?.info('Starting polling of new sessions again');
-          return;
-        } catch (e) {
-          __electronLog?.error('Failed to add token to daemons', e.message);
-          // If it fails again, just let the poller be killed
+    // We use a while loop to poll for new sessions from the client agent.
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      try {
+        sessions =
+          await this.clientAgentSessions.getNewSessionsWithCredentials();
+      } catch (e) {
+        // If we're unauthenticated, try and re-authenticate
+        if (e.statusCode === 401 || e.statusCode === 403) {
+          const sessionData = this.session.data?.authenticated;
+          const auth_token_id = sessionData?.id;
+          const token = sessionData?.token;
+
+          try {
+            await this.ipc.invoke('addTokenToDaemons', {
+              tokenId: auth_token_id,
+              token,
+            });
+
+            this.poller.perform();
```

Comment:
Can you reference this method recursively to start the task?

Reply:
yup, i tested this by throwing an error so it gets to this block. the currently running task will get canceled and new task will start running from the beginning. Also just realized that the log call should be moved before this perform call. thank u